include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h

```diff
@@ -0,0 +1,125 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_SYCL_BACKEND_IMPL_H
+#define _ONEDPL_SYCL_BACKEND_IMPL_H
+
+#include <CL/sycl.hpp>
+#include "oneapi/dpl/internal/dynamic_selection_traits.h"
+#include "oneapi/dpl/internal/dynamic_selection_impl/scoring_policy_defs.h"
+#include "oneapi/dpl/internal/dynamic_selection_impl/backend_defs.h"
+
+#include <chrono>
+#include <vector>
+#include <memory>
+
+namespace oneapi {
+namespace dpl {
+namespace experimental {
+
+  class sycl_backend {
+  public:
+    using resource_type = sycl::queue;
+    using wait_type = sycl::event;
+    using execution_resource_t = oneapi::dpl::experimental::basic_execution_resource_t<resource_type>;
+    using resource_container_t = std::vector<execution_resource_t>;
+
+  private:
+
+    class async_waiter {
+      sycl::event e_;
+      public:
+        async_waiter(sycl::event e) : e_(e) {}
+        sycl::event unwrap() { return e_; }
+        void wait() { e_.wait(); }
+    };
+
+    class submission_group {
+      resource_container_t resources_;
+    public:
+      submission_group(const resource_container_t& v) : resources_(v) { }
+
+      void wait() {
+        for (auto& r : resources_) {
+          unwrap(r).wait();
+        }
+      }
+    };
+
+  public:
+
+    sycl_backend(const sycl_backend& v) = delete;
+
+    sycl_backend() {
+      initialize_default_resources();
+      sgroup_ptr_ = std::make_unique<submission_group>(global_rank_);
+    }
+
+    template<typename NativeUniverseVector>
+    sycl_backend(const NativeUniverseVector& v) {
+      for (auto e : v) {
+        global_rank_.push_back(e);
+      }
+      sgroup_ptr_ = std::make_unique<submission_group>(global_rank_);
+    }
+
+    template<typename SelectionHandle, typename Function, typename ...Args>
+    auto submit(SelectionHandle s, Function&& f, Args&&... args) {
+      auto q = unwrap(s);
+      if constexpr (report_info_v<SelectionHandle, execution_info::task_submission_t>) {
+        report(s, execution_info::task_submission);
+      }
+      if constexpr(report_info_v<SelectionHandle, execution_info::task_completion_t>
+                   || report_value_v<SelectionHandle, execution_info::task_time_t>) {
+        std::chrono::steady_clock::time_point t0;
+        if constexpr (report_value_v<SelectionHandle, execution_info::task_time_t>) {
+          t0 = std::chrono::steady_clock::now();
+        }
+        auto e1 = f(q, std::forward<Args>(args)...);
+        auto e2 = q.submit([=](sycl::handler& h){
+            h.depends_on(e1);
+            h.host_task([=](){
+              if constexpr(report_info_v<SelectionHandle, execution_info::task_completion_t>){
+                s.report(execution_info::task_completion);
+              }
+              if constexpr(report_value_v<SelectionHandle, execution_info::task_time_t>)
+                s.report(execution_info::task_time, (std::chrono::steady_clock::now() - t0).count());
```

Comment:
I would change the order of the two checks, so that reporting task completion is outside of the measured execution time (consistent with the task submission report being also outside)